core/src/main/scala/filodb.core/memstore/IndexMetadataStore.scala

```diff
@@ -86,6 +91,249 @@ trait IndexMetadataStore {
+                                        (IndexState.Synced, Some(snapTime))
+          case Failure(e)           =>  throw e           // Failure to read the snap file in currentState is fatal
+        }
+      }.get
+    }
+  }
+
+  /**
+   *  Gets the snapTime from the snap file
+   *
+   * @param snapFile   the snapFile
+   * @param fin        The input stream object for the file
+   * @return           Long snap time if file is valid
+   */
+  private def getSnapTime(snapFile: String, fin: FileInputStream): Try[Long] = {
+        val magicByte = fin.read()
+        if (magicByte != FileSystemBasedIndexMetadataStore.snapFileV1Magic) {
+          // if first byte is not the expected magicByte then the file is not as expected and possibly garbled
+          logger.warn(" Snap {} exist and possibly corrupt, this will trigger a rebuild", snapFile.toString)
+          Failure(new IllegalStateException("Invalid magic bytes in snap file"))
+        } else {
+          // If the first byte is magicByte read just the next 8 bytes
+          val bytes = Array[Byte](0, 0, 0, 0, 0, 0, 0, 0)
+          if (fin.read(bytes) != 8) {
+            // Not able to read 8 bytes, the file is truncated and we will trigger a rebuild
+            logger.warn("Snap file {} with right magic bytes but less than 8 for time," +
+              " this will trigger a rebuild of index", snapFile.toString)
+            Failure(new IllegalStateException("Snap file truncated"))
+          } else {
+            val lastSyncTime = ByteBuffer.wrap(bytes).getLong()
+            Success(lastSyncTime)
+          }
+        }
+   }
+  /**
+   *  Updates the state of the index, the snap file will be updated with the correct timestamp in case the state
+   *  is Synced, current implementation of updateState ignores other index states. Also when the index is updated
+   *  and synced, it is important to ensure the gen file is updated with the current generation. This guarantees that
+   *  next time the on bootstrap, the initState should return Synced and not TriggerRebuild
+   *
+   * @param datasetRef The dataset ref
+   * @param shard the shard id of index
+   * @param state One of the IndexState.Values for the index
+   * @param time a time in millis since epoch for when the state was updated
+   */
+  // scalastyle:off method.length
+  override def updateState(datasetRef: DatasetRef, shard: Int, state: IndexState.Value, time: Long): Unit = {
+    state match {
+      case IndexState.Synced           =>  // Only handle cases where state is Synced
+                                    val snapFile = new File(snapFilePathFormat.format(datasetRef.dataset, shard))
+                                    val genFile = new File(genFilePathFormat.format(datasetRef.dataset, shard))
+                                    Using(new FileOutputStream(snapFile)){
+                                      // First write the last synced time to the snap file
+                                      snapFos =>
```

Comment:
separate function and UT for snapshot file write  ? ( UT would call this block and checks if it is written to file correctly )

Reply:
We have unit tests in ``FileSystemBasedIndexMetadataStoreSpec`` testing the expected functionality at higher level than reading the actual files written.